source/http_client_connection.c

```diff
@@ -0,0 +1,551 @@
+ * permissions and limitations under the License.
+ */
+#include "http_client_connection.h"
+
+#include "io.h"
+
+#include <aws/common/array_list.h>
+#include <aws/http/request_response.h>
+#include <aws/io/socket.h>
+
+#include <memoryobject.h>
+
+const char *s_capsule_name_http_client_connection = "aws_http_client_connection";
+const char *s_capsule_name_http_client_stream = "aws_http_client_stream";
+
+struct py_http_connection {
+    struct aws_allocator *allocator;
+    struct aws_http_connection *connection;
+    PyObject *capsule;
+    PyObject *on_connection_setup;
+    PyObject *on_connection_shutdown;
+    bool destructor_called;
+    bool shutdown_called;
+};
+
+static void s_http_client_connection_destructor(PyObject *http_connection_capsule) {
+    struct py_http_connection *http_connection =
+        PyCapsule_GetPointer(http_connection_capsule, s_capsule_name_http_client_connection);
+    assert(http_connection);
+    http_connection->destructor_called = true;
+    if (http_connection->connection) {
+
+        if (aws_http_connection_is_open(http_connection->connection)) {
+            aws_http_connection_close(http_connection->connection);
+        }
+
+        aws_http_connection_release(http_connection->connection);
+        http_connection->connection = NULL;
+    }
+
+    if (http_connection->shutdown_called) {
+        aws_mem_release(http_connection->allocator, http_connection);
+    }
+}
+
+static void s_on_client_connection_setup(struct aws_http_connection *connection, int error_code, void *user_data) {
+
+    struct py_http_connection *py_connection = user_data;
+    PyGILState_STATE state = PyGILState_Ensure();
+    PyObject *result = NULL;
+    PyObject *capsule = NULL;
+
+    if (!error_code) {
+        py_connection->connection = connection;
+        capsule =
+            PyCapsule_New(py_connection, s_capsule_name_http_client_connection, s_http_client_connection_destructor);
+        py_connection->capsule = capsule;
+    } else {
+        aws_mem_release(py_connection->allocator, py_connection);
+    }
+
+    result = PyObject_CallFunction(py_connection->on_connection_setup, "(Ni)", capsule, error_code);
+
+    Py_DECREF(py_connection->on_connection_setup);
+    Py_XDECREF(result);
+
+    PyGILState_Release(state);
+}
+
+static void s_on_client_connection_shutdown(struct aws_http_connection *connection, int error_code, void *user_data) {
+    (void)connection;
+    struct py_http_connection *py_connection = user_data;
+    py_connection->shutdown_called = true;
+
+    if (!py_connection->destructor_called) {
+        PyGILState_STATE state = PyGILState_Ensure();
+        PyObject *result = PyObject_CallFunction(py_connection->on_connection_shutdown, "(i)", error_code);
+        Py_XDECREF(result);
+        PyGILState_Release(state);
+    } else {
+        aws_mem_release(py_connection->allocator, py_connection);
+    }
+
+    Py_DECREF(py_connection->on_connection_shutdown);
```

Comment:
Same here, you just deleted py_connection, you need to invoke on_connection_shutdown before deletion or cache it.